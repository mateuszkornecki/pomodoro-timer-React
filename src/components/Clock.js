import React from "react";

function Clock({ className = "", hours, minutes, seconds, miliseconds }) {
    function timeInterval(min, max, value) {
        //It will return a number between min and max
        value = Math.max(min, Math.min(value, max));
        return value;
    }

    function fillWithZeros(targetLength, value) {
        //It will add zeros  before value till value.length = targetLength
        value = value.toString();
        while (value.length < targetLength) {
            value = `0${value}`;
        }
        return value;
    }

    function normalizeTime(min, max, targetLength, value) {
        return fillWithZeros(targetLength, timeInterval(min, max, value));
    }

    hours = normalizeTime(0, 24, 2, hours);
    minutes = normalizeTime(0, 59, 2, minutes >= 60 ? minutes % 60 : minutes);
    seconds = normalizeTime(0, 59, 2, seconds);
    miliseconds = normalizeTime(0, 999, 3, miliseconds);

    return (
        <h2 className={"Clock " + className}>
            {" "}
            Pozosta≈Ço {hours}:{minutes}:{seconds}:{miliseconds}
        </h2>
    );
}

export default Clock;
