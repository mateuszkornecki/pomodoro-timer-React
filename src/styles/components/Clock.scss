.clock {
    color: white;
    font-weight: normal;
    &__numbers {
        font-weight: bold;
        background: none;
        &--colored:nth-of-type(1) {
            color: $hours-color;
        }
        &--colored:nth-of-type(2) {
            color: $minutes-color;
        }
        &--colored:nth-of-type(3) {
            color: $seconds-color;
        }
        &--colored:nth-of-type(4) {
            color: $miliseconds-color;
        }
    }
}
