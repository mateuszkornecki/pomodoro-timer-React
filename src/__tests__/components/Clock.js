import React from 'react';
import Clock from '../../components/Clock';
import ReactDOM from "react-dom"
import TestRenderer from 'react-test-renderer'; // ES6


var root = null;
var clockRenderer = null;

describe('<Clock/>', () => {
    beforeEach(() => {
        root = document.createElement('div');
        ReactDOM.render(
            <Clock hours={1} minutes={20} seconds={42} miliseconds={311} />
            , root);
    });
    describe('tests for rendering', () => {
        it('render h2 when given hours, minutes , seconds and miliseconds', () => {
            expect(root.childNodes[0].nodeName).toBe("H2");
        });
        it('has className ="clock" when given hours, minutes, seconds, and miliseconds', () => {
            expect(root.childNodes[0].className).toMatch(/clock/);
        });
        it('render time properly when given hours, minutes, seconds and miliseconds', () => {
            expect(root.childNodes[0].textContent).toMatch(/01:20:42:311/);
        });

    });
    describe('sets correctly default props values', () => {
        it('sets Classname to empty string and hours,minutes, seconds, miliseconds to 0 if not given anything else', () => {
            expect(<Clock />).toEqual(<Clock className="" hours={0} minutes={0} seconds={0} miliseconds={0} />);
        });
    });
});

describe('<Clock/> with react-test-renderer', () => {
    beforeEach(() => {
        clockRenderer = TestRenderer.create(
            <Clock hours={1} minutes={20} seconds={42} miliseconds={311} />
        );
    });
    describe('tests for rendering', () => {
        it('render h2 when given hours, minutes , seconds and miliseconds', () => {
            expect(clockRenderer.toJSON().type).toEqual("h2");
        });
        //! PYTANIE NA LIVE, dlaczego nie zrobić po prostu ...props.className żeby dostać się do konkretnego klucza?
        it('has className ="clock" when given hours, minutes, seconds, and miliseconds', () => {
            expect(clockRenderer.toJSON().props.className).toMatch(/clock/);
        });
        describe('render time properly when given hours, minutes, seconds and miliseconds', () => {

            it('render hours', () => {
                expect(clockRenderer.toTree().props.hours).toBe(1);
            });
            it('render minutes', () => {
                expect(clockRenderer.toTree().props.minutes).toBe(20);
            });
            it('render seconds', () => {
                expect(clockRenderer.toTree().props.seconds).toBe(42);
            });
            it('render miliseconds', () => {
                expect(clockRenderer.toTree().props.miliseconds).toBe(311);
            });
        });

    });
    describe('sets correctly default props values', () => {
        it('sets Classname to empty string and hours,minutes, seconds, miliseconds to 0 if not given anything else', () => {
            expect(<Clock />).toEqual(<Clock className="" hours={0} minutes={0} seconds={0} miliseconds={0} />);
        });
    });
});