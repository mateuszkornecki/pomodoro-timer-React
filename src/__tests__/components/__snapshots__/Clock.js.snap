// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Clock/> with react-test-renderer tests for rendering when given hours, minutes, seconds, miliseconds renders properly 1`] = `
<h2
  className="clock "
>
   
  Pozostało 
  <mark
    className="clock__numbers"
  >
    01
  </mark>
  :
  <mark
    className="clock__numbers"
  >
    20
  </mark>
  :
  <mark
    className="clock__numbers"
  >
    42
  </mark>
  :
  <mark
    className="clock__numbers"
  >
    311
  </mark>
</h2>
`;

exports[`<Clock/> with react-test-renderer when given hours, minutes, seconds, miliseconds renders properly 1`] = `
<h2
  className="clock "
>
   
  Pozostało 
  <mark
    className="clock__numbers"
  >
    01
  </mark>
  :
  <mark
    className="clock__numbers"
  >
    20
  </mark>
  :
  <mark
    className="clock__numbers"
  >
    42
  </mark>
  :
  <mark
    className="clock__numbers"
  >
    311
  </mark>
</h2>
`;
